File: research/explorations/bench-realistic/hashmap.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"time"
)

type Entry struct{ k, v, sum int64 }

func bench(n int64) int64 {
	var s int64
	for i := int64(0); i < n; i++ {
		e := Entry{i, i * 7, 0}
		e2 := Entry{e.k, e.v, e.k + e.v}
		s += e2.sum
	}
	return s
}

func main() {
	n := int64(1000)
	if len(os.Args) > 1 {
		if v, err := strconv.ParseInt(os.Args[1], 10, 64); err == nil {
			n = v
		}
	}
	for i := 0; i < 1000; i++ {
		bench(n)
	}
	iters := int64(10000)
	start := time.Now()
	var r int64
	for i := int64(0); i < iters; i++ {
		r = bench(n)
	}
	elapsed := time.Since(start).Nanoseconds()
	fmt.Printf("result:     %d\n", r)
	fmt.Printf("iterations: %d\n", iters)
	fmt.Printf("total:      %.2fms\n", float64(elapsed)/1e6)
	fmt.Printf("per call:   %dns\n", elapsed/iters)
}

File: research/explorations/bench-realistic/hashmap.py
import sys, time
from collections import namedtuple

Entry = namedtuple('Entry', ['k', 'v', 'sum'])

def bench(n):
    s = 0
    for i in range(n):
        e = Entry(i, i * 7, 0)
        e2 = e._replace(sum=e.k + e.v)
        s += e2.sum
    return s

N = int(sys.argv[1]) if len(sys.argv) > 1 else 1000
for _ in range(100):
    bench(N)
iters = 10000
start = time.monotonic_ns()
for _ in range(iters):
    r = bench(N)
elapsed = time.monotonic_ns() - start
print(f"result:     {r}")
print(f"iterations: {iters}")
print(f"total:      {elapsed / 1e6:.2f}ms")
print(f"per call:   {elapsed // iters}ns")
